### ISSUE TITLE
Consider C# 9 top-level statements for Design-time DbContext Creation

### ISSUE THREAD

**Issue by gojanpaolo** [dotnet/efcore#23340]
I migrated to .NET 5 and tried to used top-level statement for my `Program` class. Unfortunately, EF Core's migration got broken because it was not able to detect the top-level `Program`.

It would be nice if EF Core would detect the top-level `Program`.

See also https://docs.microsoft.com/en-us/ef/core/cli/dbcontext-creation
> The tools first try to obtain the service provider by invoking `Program.CreateHostBuilder()`, calling `Build()`, then accessing the `Services` property.

**Comment by ajcvickers** [comment]
@Pilchie @mkArtakMSFT Where are host builder issues tracked these days?

**Comment by Pilchie** [comment]
In dotnet/runtime.  cc @ericstj.

**Comment by ericstj** [comment]
My understanding is that the HostFactoryResolver [code](https://github.com/dotnet/runtime/blob/6072e4d3a7a2a1493f514cdf4be75a3d56580e84/src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs#L9-L111) is intentionally rigid and only meant to match a couple patterns which are used in project templates

Can you provide a sample of what code you tried to make work here?  I think we'd need folks to agree upon what we'd like this to look like and decide if its something worth enabling.  @davidfowl / @bradygaster -- are there any plans to change the templates to use top-level functions?

Here's what I tried.
```c#
using System;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.DependencyInjection;

CreateHostBuilder(args).Build().Run();

static IHostBuilder CreateHostBuilder(string[] args) =>
    Host.CreateDefaultBuilder(args)
        .ConfigureServices((hostContext, services) =>
        {
            services.AddHostedService<Worker>();
        });

class Worker : IHostedService 
{ 
   public async Task StartAsync(CancellationToken token) => Console.WriteLine("Hello World");
   public async Task StopAsync(CancellationToken token) => Console.WriteLine("Goodbye World");
}
```

Instead it emitted `'<Program>$'::'<<Main>$>g__CreateHostBuilder|0_0'` since `CreateHostBuilder` is treated as a local function instead of a public method on the type.  As a result the HostFactoryResolver didn't find it.

Sample: [topLevel.zip](https://github.com/dotnet/efcore/files/5549974/topLevel.zip)

@maryamariyan @eerhardt

**Comment by bradygaster** [comment]
There has been no discussion to change this around to be a top-level program. Yet. I do like the syntax, however. @davidfowl what say you? I wouldn't mind this, but I'd personally prefer it if we talk to some customers and see what they think before we do it.

**Comment by ajcvickers** [comment]
@ericstj @bradygaster The feeling from the EF team when we discussed this yesterday was that it would be nice if this worked, but we should probably wait for more feedback that people actually want this before investing in it. If this were EF, then I would make this an enhancement on the backlog and ask people to vote if they are interested. Can we do the equivalent thing on dotnet/runtime?

**Comment by bradygaster** [comment]
@ajcvickers I like the idea

**Comment by ericstj** [comment]
Will it get the right attention?  The code lives there out of convenience, the consumers are EF and ASP.NETCore.  I'm fine with having this transferred to dotnet/runtime and placed in future but can EF & ASP.NETCore make whatever edits are needed to get the right eyes on it?

**Comment by davidfowl** [comment]
I agree with Eric. I'd leave it in the ASP.NET repo to get feedback.

**Comment by bradygaster** [comment]
Also agree. How should we proceed?
